src/Orleans/Streams/Internal/StreamSubscriptionChangeHandler.cs

```diff
@@ -0,0 +1,91 @@
+using Orleans.Streams.Core;
+
+namespace Orleans.Streams
+{
+    internal class StreamSubscriptionChangeHandler
+    {
+        private IStreamProviderRuntime providerRuntime;
+        private Dictionary<Type, IStreamSubscriptionObserverProxy> subscriptionObserverMap;
+
+        public StreamSubscriptionChangeHandler(IStreamProviderRuntime providerRT, Dictionary<Type, IStreamSubscriptionObserverProxy> observerMap)
+        {
+            this.providerRuntime = providerRT;
+            this.subscriptionObserverMap = observerMap;
+        }
+
+        /// <summary>
+        /// This method finds the correct IStreamSubscriptionObserver<T> based on item's type, and invoke its OnNewSubscription method
+        /// </summary>
+        /// <param name="subscriptionId"></param>
+        /// <param name="streamId"></param>
+        /// <param name="messageType"></param>
+        /// <returns></returns>
+        public async Task HandleNewSubscription(GuidId subscriptionId, StreamId streamId, Type messageType)
+        {
+            // if no observer attached to the subscription, check if the grain is a IStreamSubscriptionObserver
+            IStreamSubscriptionObserverProxy observerProxy;
+            if (TryGetStreamSubscriptionObserverProxyForType(messageType, out observerProxy))
+            {
+                var streamProvider = this.providerRuntime.ServiceProvider
+                            .GetService<IStreamProviderManager>()
+                            .GetStreamProvider(streamId.ProviderName);
+
+                await observerProxy.InvokeOnNewSubscription(streamId, subscriptionId, streamProvider);
+            }
+        }
+
+        private bool TryGetStreamSubscriptionObserverProxyForType(Type concretType, out IStreamSubscriptionObserverProxy observerProxy)
+        {
+            IStreamSubscriptionObserverProxy proxy = null;
+            foreach (var observerEntry in this.subscriptionObserverMap)
+            {
+                if (observerEntry.Key.IsAssignableFrom(concretType))
+                {
+                    proxy = observerEntry.Value;
+                    break;
+                }
+            }
+            observerProxy = proxy;
+            return proxy != null;
+        }
+    }
+
+    internal interface IStreamSubscriptionObserverProxy 
+    {
+        Task InvokeOnNewSubscription(StreamId streamId, GuidId subscriptionId, IStreamProvider streamProvider);
+        IAddressable SubscriptionObserver { set; }
+    }
+
+    /// <summary>
+    /// Decorator class for a IStreamSubscriptionObserver<T>. Created mainly to avoid using reflection to invoke
+    /// methods on IStreamSubscriptionObserver<T>.
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    internal class StreamSubscriptionObserverProxy<T> : IStreamSubscriptionObserver<T>, IStreamSubscriptionObserverProxy
+    {
+        private IStreamSubscriptionObserver<T> observer => this.SubscriptionObserver as IStreamSubscriptionObserver<T>;
```

Comment:
suggest a hard cast on the property setter rather than an 'as' on each read.  If someone uses this wrong I think a hard exception as early as possible is better than a silent null value that could surface as a null reference exception far from the cause of the problem.